## a. Download dataset

In [34]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re

In [35]:
data=pd.read_csv("input/Dollar_Dataset.csv")
data.drop(columns = 'Unnamed: 0', inplace = True)
data.shape

(4214, 6)

In [36]:
data[data.duplicated()]
data.drop_duplicates(inplace=True)
data['High'] = data['High'].apply(lambda x: re.sub(',', '', x))
data["High"] = pd.to_numeric(data["High"] , errors='coerce')
data['Low'] = data['Low'].apply(lambda x: re.sub(',', '', x))
data["Low"] = pd.to_numeric(data["Low"] , errors='coerce')
data['Open'] = data['Open'].apply(lambda x: re.sub(',', '', x))
data["Open"] = pd.to_numeric(data["Open"] , errors='coerce')
data['Close'] = data['Close'].apply(lambda x: re.sub(',', '', x))
data["Close"] = pd.to_numeric(data["Close"] , errors='coerce')
data.drop(columns = 'Persian_Date', inplace = True)

print(data.shape)
data.head()

(3269, 5)


Date    Open     Low    High   Close
0  2023/07/26  491550  488700  491600  488990
1  2023/07/25  484470  483900  488900  488370
2  2023/07/24  480700  480100  486400  486140
3  2023/07/23  485060  481600  485400  483360
4  2023/07/22  487830  485600  487900  485810

## b. Divide dataset to Ahmadinejad, Rouhani and Raisi's presidency

In [37]:
data['Date'] = pd.to_datetime(data['Date'], format="%Y/%m/%d")
ahmadinejad=data.loc[(data['Date'] >= "2011/11/27")
                     & (data['Date'] < "2013/08/03")].reset_index()
rouhani=data.loc[(data['Date'] >= "2013/08/03")
                     & (data['Date'] < "2021/08/03")].reset_index()
raisi=data.loc[(data['Date'] >= "2021/08/03")
                     & (data['Date'] < "2023/07/26")].reset_index()
print(ahmadinejad.shape)
print(rouhani.shape)
print(raisi.shape)
rouhani.head()

(482, 6)
(2316, 6)
(470, 6)


index       Date    Open     Low    High   Close
0    721 2021-08-02  255690  254590  256640  255190
1    722 2021-08-01  257190  254290  257640  254890
2    723 2021-07-31  260270  256990  260640  257290
3    724 2021-07-28  251800  251690  256240  255990
4    725 2021-07-27  250780  250690  252340  251660

## c. Show the highest dollar price in Ahmadinejad, Rouhani and Raisi's presidency respectively

In [38]:
max_price_dollar_ahmadinejad=ahmadinejad['High'].max()
print(f'max price of dollar in ahmadinejad presidency: {max_price_dollar_ahmadinejad}')
max_price_dollar_rouhani=rouhani['High'].max()
print(f'max price of dollar in Rouhani presidency: {max_price_dollar_rouhani}')
max_price_dollar_raisi=raisi['High'].max()
print(f'max price of dollar in raisi presidency: {max_price_dollar_raisi}')

max price of dollar in ahmadinejad presidency: 39700
max price of dollar in Rouhani presidency: 320060
max price of dollar in raisi presidency: 555600


|  ------------Max Dollar Price In 3 presidency----------------        |
|-------------------------------------------------------------------------|

|                  |     Price(Rial)    |
|------------------|------------|
|      max price of dollar in ahmadinejad presidency      | 39700 | 
|      max price of dollar in Rouhani presidency          | 320060 |
|      max price of dollar in raisi presidency            | 555600 |

## d. Show the lowest dollar price in Ahmadinejad, Rouhani and Raisi's presidency respectively

In [39]:
max_price_dollar_ahmadinejad=ahmadinejad['Low'].min()
print(f'max price of dollar in ahmadinejad presidency: {max_price_dollar_ahmadinejad}')
max_price_dollar_rouhani=rouhani['Low'].min()
print(f'max price of dollar in Rouhani presidency: {max_price_dollar_rouhani}')
max_price_dollar_raisi=raisi['Low'].min()
print(f'max price of dollar in raisi presidency: {max_price_dollar_raisi}')

max price of dollar in ahmadinejad presidency: 13350
max price of dollar in Rouhani presidency: 12850
max price of dollar in raisi presidency: 251250


|  -------------Min Dollar Price In 3 presidency---------------        |
|-------------------------------------------------------------------------|

|                  |     Price(Rial)    |
|------------------|------------|
|      min price of dollar in ahmadinejad presidency      | 13350 | 
|      min price of dollar in Rouhani presidency          | 12850 |
|      min price of dollar in raisi presidency            | 251250 |

## E. Split each dataset to train and test with train_test_split function

In [40]:
ahmadinejad.corr()

index      Date      Open       Low      High     Close
index  1.000000 -0.999471 -0.906595 -0.906979 -0.903370 -0.903989
Date  -0.999471  1.000000  0.905611  0.905984  0.902603  0.903232
Open  -0.906595  0.905611  1.000000  0.999067  0.998685  0.998398
Low   -0.906979  0.905984  0.999067  1.000000  0.997781  0.998602
High  -0.903370  0.902603  0.998685  0.997781  1.000000  0.999507
Close -0.903989  0.903232  0.998398  0.998602  0.999507  1.000000

In [41]:
from sklearn.model_selection import train_test_split

ahmadinejad['Date']=ahmadinejad['Date'].dt.strftime('%Y%m%d')
rouhani['Date']=rouhani['Date'].dt.strftime('%Y%m%d')
raisi['Date']=raisi['Date'].dt.strftime('%Y%m%d')

ahmadinejad["Date"] = pd.to_numeric(ahmadinejad["Date"] , errors='coerce')
rouhani["Date"] = pd.to_numeric(rouhani["Date"] , errors='coerce')
raisi["Date"] = pd.to_numeric(raisi["Date"] , errors='coerce')

X_ahmadinejad=ahmadinejad['Date'].values
Y_ahmadinejad=ahmadinejad['High'].values

X_rouhani=rouhani['Date'].values
Y_rouhani=rouhani['High'].values

X_raisi=raisi['Date'].values
Y_raisi=raisi['High'].values

X_train_ahmadinejad, X_test_ahmadinejad,Y_train_ahmadinejad, Y_test_ahmadinejad=train_test_split(X_ahmadinejad,Y_ahmadinejad,test_size=0.2)
X_train_rouhani, X_test_rouhani,Y_train_rouhani, Y_test_rouhani=train_test_split(X_rouhani,Y_rouhani,test_size=0.2)
X_train_raisi, X_test_raisi,Y_train_raisi, Y_test_raisi=train_test_split(X_raisi,Y_raisi,test_size=0.2)

X_train_ahmadinejad=X_train_ahmadinejad.reshape(-1,1)
X_test_ahmadinejad=X_test_ahmadinejad.reshape(-1,1)
Y_train_ahmadinejad=Y_train_ahmadinejad.reshape(-1,1)
Y_test_ahmadinejad=Y_test_ahmadinejad.reshape(-1,1)

X_train_rouhani=X_train_rouhani.reshape(-1,1)
X_test_rouhani=X_test_rouhani.reshape(-1,1)
Y_train_rouhani=Y_train_rouhani.reshape(-1,1)
Y_test_rouhani=Y_test_rouhani.reshape(-1,1)

X_train_raisi=X_train_raisi.reshape(-1,1)
X_test_raisi=X_test_raisi.reshape(-1,1)
Y_train_raisi=Y_train_raisi.reshape(-1,1)
Y_test_raisi=Y_test_raisi.reshape(-1,1)

print(f"shape of x train ahmadinejad: {X_train_ahmadinejad.shape}")
print(f"shape of y train ahmadinejad: {Y_train_ahmadinejad.shape}")
print(f"shape of x test ahmadinejad: {X_test_ahmadinejad.shape}")
print(f"shape of y test ahmadinejad: {Y_test_ahmadinejad.shape}")
print()
print(f"shape of x train rouhani: {X_train_rouhani.shape}")
print(f"shape of y train rouhani: {Y_train_rouhani.shape}")
print(f"shape of x test rouhani: {X_test_rouhani.shape}")
print(f"shape of y test rouhani: {Y_test_rouhani.shape}")
print()
print(f"shape of x train raisi: {X_train_raisi.shape}")
print(f"shape of y train raisi: {Y_train_raisi.shape}")
print(f"shape of x test raisi: {X_test_raisi.shape}")
print(f"shape of y test raisi: {Y_test_raisi.shape}")

shape of x train ahmadinejad: (385, 1)
shape of y train ahmadinejad: (385, 1)
shape of x test ahmadinejad: (97, 1)
shape of y test ahmadinejad: (97, 1)

shape of x train rouhani: (1852, 1)
shape of y train rouhani: (1852, 1)
shape of x test rouhani: (464, 1)
shape of y test rouhani: (464, 1)

shape of x train raisi: (376, 1)
shape of y train raisi: (376, 1)
shape of x test raisi: (94, 1)
shape of y test raisi: (94, 1)


## F. Fit the LLS model on each training dataset

In [43]:
from lls import LinearLeastSquare

my_lls_ahmadinejad= LinearLeastSquare()
my_lls_rouhani= LinearLeastSquare()
my_lls_raisi= LinearLeastSquare()
my_lls_ahmadinejad.fit(X_train_ahmadinejad,Y_train_ahmadinejad)
my_lls_rouhani.fit(X_train_rouhani,Y_train_rouhani)
my_lls_raisi.fit(X_train_raisi,Y_train_raisi)

Y_pred_ahmadinejad=my_lls_ahmadinejad.predict(X_test_ahmadinejad)
Y_pred_rouhani=my_lls_rouhani.predict(X_test_rouhani)
Y_pred_raisi=my_lls_raisi.predict(X_test_raisi)

## G. Evaluate each model on test dataset using MAE loss function in Scikit-Learn

In [46]:
from sklearn.metrics import mean_absolute_error


mae_ahmadinejad=mean_absolute_error(Y_test_ahmadinejad,Y_pred_ahmadinejad)
mae_rouhani=mean_absolute_error(Y_test_rouhani,Y_pred_rouhani)
mae_raisi=mean_absolute_error(Y_test_raisi,Y_pred_raisi)

print(f'MAE for ahmadinejad presidency: {mae_ahmadinejad}')
print(f'MAE for rouhani presidency: {mae_rouhani}')
print(f'MAE for raisi presidency: {mae_raisi}')

MAE for ahmadinejad presidency: 8058.283383502259
MAE for rouhani presidency: 62206.575309224754
MAE for raisi presidency: 81580.80689561363


| MAE for Dollar Price Prediction In 3 presidency       |
|-------------------------------------------------------------------------|

|                  |     Price(Rial)    |
|------------------|------------|
|      MAE for ahmadinejad presidency      | 8058.283383 | 
|      MAE for rouhani presidency          | 62206.575309 |
|      MAE for raisi presidency            | 81580.806895 |